### ISSUE TITLE
unable to run example script on Windows 10

### ISSUE THREAD

**Issue by EvoArt** [StanJulia/StanSample.jl#42]
First off, thanks for making this package. It's great (or it will be once I've got it working) to be able to combine Stan with Julia.

I receive the following error when trying to call ```m6_1s = SampleModel("m6.1s", stan6_1);```
```julia
C:\Users\arn203\AppData\Local\Temp\jl_LcNS1c\m6.1s.stan updated.
ERROR: IOError: could not spawn `make -f C:/.cmdstan/cmdstan-2.27.0/makefile -C C:/.cmdstan/cmdstan-2.27.0/ 'C:\Users\arn203\AppData\Local\Temp\jl_LcNS1c\m6.1s.exe'`: no such file or directory (ENOENT)
Stacktrace:
  [1] _spawn_primitive(file::String, cmd::Cmd, stdio::Vector{Any})
    @ Base .\process.jl:99
  [2] #637
    @ .\process.jl:112 [inlined]
  [3] setup_stdios(f::Base.var"#637#638"{Cmd}, stdios::Vector{Any})
    @ Base .\process.jl:196
  [4] _spawn
    @ .\process.jl:111 [inlined]
  [5] _spawn(::Base.CmdRedirect, ::Vector{Any})
    @ Base .\process.jl:139
  [6] _spawn
    @ .\process.jl:106 [inlined]
  [7] success(cmd::Base.CmdRedirect)
    @ Base .\process.jl:496
  [8] #29
    @ C:\Users\arn203\.julia\packages\StanSample\en13m\src\stanmodel\SampleModel.jl:57 [inlined]
  [9] cd(f::WARNING: both AxisKeys and DataFrames export "rename"; uses of it in module StanSample must be qualified
StanSample.var"#29#30"{IOBuffer, String, String}, dir::String)
    @ Base.Filesystem .\file.jl:95
 [10] SampleModel(name::String, model::String, n_chains::Vector{Int64}; seed::StanBase.RandomSeed, init::StanBase.Init, output::StanBase.Output, tmpdir::String, summary::Bool, printsummary::Bool, method::StanSample.Sample)
    @ StanSample C:\Users\arn203\.julia\packages\StanSample\en13m\src\stanmodel\SampleModel.jl:56
 [11] SampleModel (repeats 2 times)
    @ C:\Users\arn203\.julia\packages\StanSample\en13m\src\stanmodel\SampleModel.jl:47 [inlined]
 [12] top-level scope
    @ REPL[9]:1
```

I installed cmdstan using R/Rtools. Copying the path to the missing exe and changing the extension to .stan e.g. "C:\Users\arn203\AppData\Local\Temp\jl_LcNS1c\m6.1s.stan" I can build the executable from cmdstanR. But I'd much prefer to use Julia!

**Comment by goedman** [comment]
@EvoArt Thanks for the message. I'm assuming you use Julia 1.6.2? Which StanSample version, the latest published?

Did you install cmdstan or cmdstanR? I've never used the latter, but cmdstan + the Bernoulli example test script has always worked. I'll investigate where the warning about AxisKeys and DatafFrames comes from, I've never seen that one.

Rob

**Comment by EvoArt** [comment]
Thanks for getting back to me.
I'm using, Julia 1.6.1 (could that be the issue?) and  StanSample v4.0.0. 
I installled cmdstan via Rtools as I'm still figuring out how to do things on Windows. I will have a go at installing it "myself" using ```make``` and see if that makes a difference.

**Comment by goedman** [comment]
I have a hard time to reproduce your issue. Which script are you trying to execute?

I am still concerned about the warning on line 9 and it points to where above error is created in SampleModel. But that could happen because you are calling a cmdstanR makefile.

Is there a build and/or run log file in the Temp directory?

After you build cmdstan with RTools, you should be able to test with:
```
    1. Build the model:
       > make examples/bernoulli/bernoulli
    2. Run the model:
       > examples/bernoulli/bernoulli sample data file=examples/bernoulli/bernoulli.data.R
    3. Look at the samples:
       > bin/stansummary output.csv
```

**Comment by goedman** [comment]
A good test script might be:
```
######### StanSample Bernoulli example  ###########

using StanSample

ProjDir = @__DIR__

bernoulli_model = "
data {
  int<lower=1> N;
  int<lower=0,upper=1> y[N];
}
parameters {
  real<lower=0,upper=1> theta;
}
model {
  theta ~ beta(1,1);
  y ~ bernoulli(theta);
}
";

bernoulli_data = Dict("N" => 10, "y" => [0, 1, 0, 1, 0, 0, 0, 0, 0, 1])

# Keep tmpdir across multiple runs to prevent re-compilation
tmpdir = joinpath(@__DIR__, "tmp")

sm = SampleModel("bernoulli", bernoulli_model;
  method = StanSample.Sample(
    save_warmup=false,                           # Default
    thin=1,
    adapt = StanSample.Adapt(delta = 0.85)),
  tmpdir = tmpdir,
);

rc = stan_sample(sm; data=bernoulli_data);

if success(rc)
  chns = read_samples(sm)
  display(chns)
end
```

This will make it easier to compare results.

**Comment by goedman** [comment]
One last thing, can you try:
```
g++ --version
mingw32-make --version
make --version
```
on your system? Just to see in make.exe is available.

**Comment by EvoArt** [comment]
> One last thing, can you try:
> 
> ```
> g++ --version
> mingw32-make --version
> make --version
> ```
> 
> on your system? Just to see in make.exe is available.

hmm. of these, only ```make --version``` works

GNU Make 4.3
Built for Windows32
Copyright (C) 1988-2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

**Comment by EvoArt** [comment]
> This will make it easier to compare results.

Running this script I get 

```julia
julia> sm = SampleModel("bernoulli", bernoulli_model;
         method = StanSample.Sample(
             save_warmup=false,                           # Default
           thin=1,
           adapt = StanSample.Adapt(delta = 0.85)),
         tmpdir = tmpdir,
       );

C:\Users\arn203\OneDrive - University of Exeter\Documents\carbon_sources\Bead_MS\Julia_Beads\tmp\bernoulli.stan 
updated.
ERROR: IOError: could not spawn `make -f C:/.cmdstan/cmdstan-2.27.0/makefile -C C:/.cmdstan/cmdstan-2.27.0 'C:\Users\arn203\OneDrive - University of Exeter\Documents\carbon_sources\Bead_MS\Julia_Beads\tmp\bernoulli.exe'`: no such file or directory (ENOENT)
Stacktrace:
  [1] _spawn_primitive(file::String, cmd::Cmd, stdio::Vector{Any})
    @ Base .\process.jl:99
  [2] #637
    @ .\process.jl:112 [inlined]
  [3] setup_stdios(f::Base.var"#637#638"{Cmd}, stdios::Vector{Any})
    @ Base .\process.jl:196
  [4] _spawn
    @ .\process.jl:111 [inlined]
  [5] _spawn(::Base.CmdRedirect, ::Vector{Any})
    @ Base .\process.jl:139
  [6] _spawn
    @ .\process.jl:106 [inlined]
  [7] success(cmd::Base.CmdRedirect)
    @ Base .\process.jl:496
  [8] #29
    @ C:\Users\arn203\.julia\packages\StanSample\en13m\src\stanmodel\SampleModel.jl:57 [inlined]
  [9] cd(f::WARNING: both AxisKeys and DataFrames export "rename"; uses of it in module StanSample must be qualified
StanSample.var"#29#30"{IOBuffer, String, String}, dir::String)
    @ Base.Filesystem .\file.jl:95
 [10] SampleModel(name::String, model::String, n_chains::Vector{Int64}; seed::StanBase.RandomSeed, init::StanBase.Init, output::StanBase.Output, tmpdir::String, summary::Bool, printsummary::Bool, method::StanSample.Sample)  
    @ StanSample C:\Users\arn203\.julia\packages\StanSample\en13m\src\stanmodel\SampleModel.jl:56
 [11] top-level scope
    @ REPL[9]:1

```

**Comment by EvoArt** [comment]
> Looking through the RTools installation docs, I noticed:
> 
> ```
> On Windows run the following:
> 
> # compile the example
> > mingw32-make examples/bernoulli/bernoulli.exe 
> 
> # fit to provided data (results of 10 trials, 2 out of 10 successes)
> > ./examples/bernoulli/bernoulli.exe sample data file=examples/bernoulli/bernoulli.data.json
> 
> # run the `bin/stansummary.exe utility to summarize parameter estimates
> > bin/stansummary.exe output.csv
> ```
> 
> which is not what I'm using in the 2nd line above, I call `make` vs. `ming32-make`. If that is due to installing with RTools I need to enable it.

Trying both ways I get
```
PS C:\.cmdstan\cmdstan-2.27.0> make examples/bernoulli/bernoulli
INFO: Could not find files for the given pattern(s).
make: *** No rule to make target 'examples/bernoulli/bernoulli'.  Stop.
PS C:\.cmdstan\cmdstan-2.27.0> mingw32-make examples/bernoulli/bernoulli
mingw32-make : The term 'mingw32-make' is not recognized as the name of a cmdlet, function, script file, or operable
program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
At line:1 char:1
+ mingw32-make examples/bernoulli/bernoulli
+ ~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (mingw32-make:String) [], CommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException

```

So clearly the problem is not with. StannSample.jl. 

I'm going to close this for now since the problem is with my cmdstan installation. And Ill come back if there's still issues after re-installing.

**Comment by goedman** [comment]
Arthur (@EvoArt), just 2 other hints I've seen creating issues in the past. On Windows, most needed/ended up using PowerShell. The 2nd issue is blanks in file names, so above `OneDrive - University of Exeter\Documents` might not be handled well by cmdstan.

**Comment by goedman** [comment]
Also, in the first attempted line, is should be:
```
PS C:\.cmdstan\cmdstan-2.27.0> make examples/bernoulli/bernoulli.exe
```
On Windows, the `.exe` is needed.

It looks like RTools installs/renames `ming32-make` to just `make`.

**Comment by EvoArt** [comment]
>PS C:\.cmdstan\cmdstan-2.27.0> make examples/bernoulli/bernoulli.exe

Thanks. Running this line, I'm now getting:

```'cut' is not recognized as an internal or external command,```

**Comment by goedman** [comment]
Aah, that is explained here:
```
1.6.1 Common problems

This section contains solutions to problems reported on https://discourse.mc-stan.org

Compiler error message about PCH file

To speed up compilation, the Stan makefile pre-compiles parts of the core Stan library. If these pre-compiled files are out of sync with the compiled model, the compiler will complain, e.g.:

error: PCH file uses an older PCH format that is no longer supported
In this case, clean and rebuild CmdStan, as shown in the previous section.

Windows: ‘mingw32-make’ is not recognised

If the C++ toolchain has been installed but not properly registered, then the call to mingw32-make will result in error message:

'mingw32-make' is not recognised as an internal or external command
To fix this, you should make sure that the RTools installation is in your PATH environment variable, as described in the RTools section.

In addition, if you have the RTools 4.0 toolchain, then call the utility pacman to register the right version of make:

pacman -Sy mingw-w64-x86_64-make
See: https://discourse.mc-stan.org/t/cmdstan-installation-on-windows/11287/6

Windows: ‘g++’ or ‘cut’ is not recognized

The CmdStan makefile uses a few shell utilities which might not be present in Windows, resulting in the error message:

'cut' is not recognized as an internal or external command, operable program or batch file.
The solution is to add mingw_64/bin directory to the user PATH environment variable. See: https://discourse.mc-stan.org/t/errors-when-using-cmdstan-model/14984/21

```

**Comment by EvoArt** [comment]
Thanks Rob. I had already come across this and tried the recommended advice, to no avail. Funnily enough, I could atleady use cmdstan via cmdstanR before. But since installing make and g++ separately, I can't even do that anymore. Going to try and remove everything and start from scratch.

**Comment by EvoArt** [comment]
Just to update, it is in fact working now! To be honest I changed around so many settings, it's hard to know what I had to do, in what order.

Installing everything via Rtools/cmdstanR got everything working for using cmdstan from the command line and R, but I had errors from StanSample.jl.

Then installing g++ and make separately, changed the errors I got and also seemed to break cmdstanR. 

I rebuilt cmdstan using ```rebuild_cmdstan()``` in cmdstanR

In my path I have: 
    C:\rtools40\usr
    C:\rtools40\usr\bin
    C:\rtools40\mingw64\bin
    C:\Users\.....\.cmdstan\cmdstan-2.27.0\stan\lib\stan_math\lib\tbb
    
And now everything seems to work from Julia, command line and R.

Thanks for your patience.
